src/python/Utils/ExecuteDAO.py

```diff
@@ -0,0 +1,214 @@
+                        Simulation mode only""")
+    parser.add_argument('-s', '--sqlKwArgs', default={},
+                        help="""\
+                        Named paramaters to be forwarded to the DAO execute method and used as SQL arguments in the query.
+                        Should be formatted as a dictionary e.g.:
+                        -s "{'workflowName': name, injected: True}"
+                        """)
+    parser.add_argument('sqlArgs', nargs=argparse.REMAINDER, default=(),
+                        help="""\
+                        -- Positional parameters to be forwarded to the DAO execute method and used as SQL arguments in the query.""")
+
+    args = parser.parse_args()
+
+    return args
+
+
+def loggerSetup(logLevel=logging.INFO):
+    """
+    Return a logger which writes everything to stdout.
+    """
+    logger = logging.getLogger()
+    outHandler = logging.StreamHandler(sys.stdout)
+    outHandler.setFormatter(logging.Formatter("%(asctime)s:%(levelname)s:%(module)s: %(message)s"))
+    outHandler.setLevel(logLevel)
+    if logger.handlers:
+        logger.handlers.clear()
+    logger.addHandler(outHandler)
+    logger.setLevel(logLevel)
+    return logger
+
+
+class ExecuteDAO():
+    """
+    A generic class to create the DAO Factory and execute the DAO module.
+    """
+    def __init__(self, logger, package=None, daoModule=None):
+        """
+        __init__
+        The ExecuteDAO constructor method.
+        :param logger: The logger instance.
+        :param package: The Package from which the DAO factory to be initialised.
+        :param module: The DAO module to be executed.
+        """
+        # Connecting to DB
+        myThread = threading.currentThread()
+        connectToDB()
+        self.logger = logger
+        self.dbi = myThread.dbi
+        self.package = package
+        self.daoModule = daoModule
+
+        self.daoFactory = DAOFactory(package=package,
+                                     logger=self.logger,
+                                     dbinterface=self.dbi)
+        self.logger.info("DAO Factory initialised from package: %s", self.package)
+
+        self.dao = self.daoFactory(classname=daoModule)
+        self.logger.info("DAO Module initialised as: %s", self.daoModule)
+
+    def __call__(self, dryRun=False, *sqlArgs, **sqlKwArgs):
+        """
+        __call__
+        The ExecuteDAO call method. This is the method to forward all provided
+        arguments to the execute method of the DAO and return the result from the query
+        :param dryRun:      Bool flag to indicate dryrun method
+        :param *sqlArgs:    All positional arguments to be forwarded to the DAO's execute method.
+        :param **sqlKwArgs: All named arguments to be forwarded to the DAO's execute method.
+        :return:            The result from the DAO (could be either a dict or a list depending on the DAO itself)
+        """
+        if dryRun:
+            results = None
```

Comment:
@todor-ivanov a typical Python error, here you return results of type None, but on line on 142 it is either list of generator (depending on type of `dao.execute` call). It is very bad practice to have mixed types. Please identify which data type should be returned, e.g. list, document it, and make consistent return of such type. If there is no results you should return the same type, e.g. empty list, or throw exception. 

In a future, we should probably start adopting type hints, see https://docs.python.org/3/library/typing.html

Reply:
Thanks @vkuznet , you are correct we should be consistent with the data types returned. The problem is, really there is no standardized data type returned from the DAOs. It could be actually one of the following three:
* generator
* list
* dict
And the module here knowing nothing about what the DAO is about to do and return at runtime prevents any type  conversion effort at this stage.  For the current case I can return an empty list though.